### PR TITLE
trunner: armv7a9: change way that plo is loaded to board

### PR DESCRIPTION
**Pull request by maska989**
## Description
Changing way that plo is loaded to board using onboard jtag chip.

## Motivation and Context
JIRA: CI-364

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [X] Tested by hand on: armv7a9-zynq7000-zedboard.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing linter checks and tests passed.
- [X] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [X] This PR needs additional PRs to work 
- https://github.com/phoenix-rtos/phoenix-rtos-build/pull/166

- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: trunner/target/armv7a9.py
Comment (nalajcie) [inline trunner/target/armv7a9.py:44]: IMHO don't change the class name, it still can only load plo.

File: trunner/target/armv7a9.py
Comment (nalajcie) [inline trunner/target/armv7a9.py:59]: why change of encoding? Seems like trying to hide some other errors...  (as charmap-based encoding will accept any byte, will just fail at a later time - when trying to decode into UTF-8).

Also - I'm not sure it's valid without explicitly stating some charmap (like `cp1252`) - do we have non-standard locale somewhere resulting in national chars being printed?

File: trunner/tools/gdb.py
Comment (nalajcie) [inline trunner/tools/gdb.py:109]: IMHO we should not make special cases based on target here - use generic options instead, eg. extra_params or per-feature flag, eg.

possible approach with extra_args - to be defined on per-target basis in particular target implementation

```python
    def __init__(self, interface: str, target: str, extra_args: Optional[List[str]] = None):
        self.target = target
        self.interface = interface
        self.extra_args = extra_args
        # (...)

    def run(self):
        args = [
            "-f",
            f"interface/{self.interface}.cfg",
            "-f",
            f"target/{self.target}.cfg",
        ]

        if self.extra_args:
            args.extend(self.extra_args)
       
        # (...)

```

File: trunner/tools/gdb.py
Comment (nalajcie) [inline trunner/tools/gdb.py:118]: fishy encoding

**Review by nalajcie — CHANGES_REQUESTED** [review]

**Comment by badochov** [comment]
in order to test it do i need to just connect all 3 usb ports on zedboard to pc and run campaign or some additional setup is needed?

**Comment by maska989** [comment]
> in order to test it do i need to just connect all 3 usb ports on zedboard to pc and run campaign or some additional setup is needed?

To test this change you will need to connect UART(taged as J14) and Jtag(taged as PROG) USB's and setup signal on mio (from 2 to 6) jumpers to gnd.
In armv7a9.py you will need to comment out lines 40 and 41.
after that simply run test runner. You will see something similar to this:
image

File: trunner/target/armv7a9.py
Comment (damianloew) [reply trunner/target/armv7a9.py:59]: > why change of encoding? Seems like trying to hide some other errors... (as charmap-based encoding will accept any byte, will just fail at a later time - when trying to decode into UTF-8).
> 
> Also - I'm not sure it's valid without explicitly stating some charmap (like `cp1252`) - do we have non-standard locale somewhere resulting in national chars being printed?

Yes, that's not perfect, but I'd opt for leaving `charmap`, or rather `cp1252` to provide the full output - that unfortunately contains non-utf8 char. From what I see it's only used in case of an errror and then it's appended to an exception (this specific char is transformed, but that's not problematic in my opinion):
Screenshot from 2023-10-18 11-40-00

What is the exact case, where it may cause some problems?

Removing `.decode("ascii")` from `raise FlashError(msg=str(e), output=e.stdout.decode("ascii") if e.stdout else None) from e` would also be needed.

File: trunner/target/armv7a9.py
Comment (damianloew) [inline trunner/target/armv7a9.py:50]: just more readable

```suggestion
        self.gdbserver = OpenocdGdbServer(interface="ftdi/digilent_jtag_smt2", target= "zynq_7000")
```

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:103]: I am not sure `if self.extra_args` it's working, but maybe we can extend arguments even with empty list (then if won't be needed)

```suggestion
        self.extra_args = []
```

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:118]:
```suggestion
            args.extend(self.extra_args)
```

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:97]:
```suggestion
    def __init__(self, interface: str, target: str):
```

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:5]:
```suggestion
from typing import Union
```

File: trunner/target/armv7m4.py
Comment (damianloew) [inline trunner/target/armv7m4.py:49]:
```suggestion
        # Set of additional config parameters used in Openocd to flash up stm32l4a6
        self.extra_args = ["-c", "reset_config srst_only srst_nogate connect_assert_srst", "-c", "init;reset"]
```

File: trunner/target/armv7m4.py
Comment (damianloew) [inline trunner/target/armv7m4.py:54]:
```suggestion
        with OpenocdGdbServer(interface="stlink", target="stm32l4x").run():
```

**Review by damianloew — CHANGES_REQUESTED** [review]
Thanks for the research and suggesting new approach. Some potential simplifications left.

File: trunner/target/armv7m4.py
Comment (nalajcie) [reply trunner/target/armv7m4.py:49]: whaaat?  This is different `self`, please provide the extra args correctly (via a constructor to different class as it's now)

File: trunner/tools/gdb.py
Comment (nalajcie) [reply trunner/tools/gdb.py:103]: It's working, `None` and `[]` (empty list) evaluates to False in python

File: trunner/target/armv7a9.py
Comment (nalajcie) [reply trunner/target/armv7a9.py:59]: Soooo, this hides some underlying, not diagnosed problem without any comment.

IMHO we should try to pin-point (and hopefully fix) underlying issue instead of hiding it like that.

- does it happen also in non-batch (interactive) mode?
- which command results in erroneous char being printed?
- is there a global .gdbinit which may change it?
- is it related to reset not being configured?
- does gdb version change the output (eg. using `arm-none-eabi-gdb` from GCC toolchain?)

Comment (damianloew) [reply trunner/target/armv7a9.py:59]: Yeah, for sure the comment was missing. Thanks for suggestions, you are right, it's worth to check. @maska989 could you follow them?

File: trunner/tools/gdb.py
Comment (damianloew) [reply trunner/tools/gdb.py:103]: There was a mistake in my comment, now it's corrected.

File: trunner/target/armv7m4.py
Comment (damianloew) [reply trunner/target/armv7m4.py:49]: Ooh right, I confused `STM32L4x6OpenocdGdbServerHarness` with `STM32L4x6OpenocdGdbServer`, I will remove wrong suggestions.

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:97]: Then `if self.extra_args:` shouldn't be needed.

```suggestion
    def __init__(self, interface: str, target: str, extra_args: Optional[List[str]] = []):
```

File: trunner/tools/gdb.py
Comment (damianloew) [inline trunner/tools/gdb.py:5]:
```suggestion
from typing import Optional, Union, List
```

File: trunner/tools/gdb.py
Comment (nalajcie) [reply trunner/tools/gdb.py:97]: no.
https://www.codiga.io/blog/python-avoid-empty-array-parameter/

Comment (damianloew) [reply trunner/tools/gdb.py:97]: That's right - I forgot about this issue, so we can leave this as it was.

File: trunner/target/armv7a9.py
Comment (maska989) [reply trunner/target/armv7a9.py]: Problem solved via setting up reset_config and tweeks into upload-zynq7000-smt2.gdb.

Comment (nalajcie) [reply trunner/target/armv7a9.py]: great to hear that :)

File: trunner/tools/gdb.py
Comment (nalajcie) [inline trunner/tools/gdb.py:107]: right now board  and target/inferface are mutually exclusive (and changed to optional in typing), we might want to add checks for that, eg:

```python
if board:
    if target or interface:
        raise OpenocdError("target or interface arguments provided when board is specified")
else:
   if not target or not interface:
        raise OpenocdError("target or interface arguments missing")
```

probably adding docstring regarding this new semantics would also be helpful

**Review by nalajcie — COMMENTED** [review]
the code looks mych better now, one more thing to suggest is to enforce / describe new semantics for instantiation `OpenocdGdbServer` to avoid problems in the future. Otherwise LGTM (if it works ;))

**Comment by maska989** [comment]
> the code looks mych better now, one more thing to suggest is to enforce / describe new semantics for instantiation `OpenocdGdbServer` to avoid problems in the future. Otherwise LGTM (if it works ;))

garuda

Thanks to @badochov i tried new version of upload-zynq7000-smt2.gdb, and it works like a charm on his environment.

**Review by nalajcie — APPROVED** [review]

File: trunner/target/armv7a9.py
Comment (badochov) [inline trunner/target/armv7a9.py:92]: I feel that with this change `class ARMv7A9Target` loses it's previous meaning, as now it only gives support for ARMv7A9 targets that have smt2 controller on board, eg I doubt this class could be reused to test zturn target. Said that I think that class structure needs rethinking

File: trunner/target/armv7a9.py
Comment (badochov) [inline trunner/target/armv7a9.py:50]: This change further limited targets supported by this class to zedboard only thus I feel it should be renamed to `ZynqZedboardGdbPloLoader`

Comment (damianloew) [reply trunner/target/armv7a9.py:50]: I think we could rename it to avoid misunderstandings.

File: trunner/target/armv7a9.py
Comment (damianloew) [reply trunner/target/armv7a9.py]: You are right, it can may be misleading. That's why I'd suggest leaving there:

```
    @abstract_method
    def flash_dut(self):
        # for now there is no common flashing approach for all future armv7a9 targets

```

and move the valid flash_dut() to Zynq7000ZedboardTarget.

@badochov what do you think?
@maska989 could you check this solution?

Comment (badochov) [reply trunner/target/armv7a9.py]: Seems like a good solution to me

Comment (badochov) [reply trunner/target/armv7a9.py]: I'll approve PR once this change is made

Comment (maska989) [reply trunner/target/armv7a9.py]: @badochov done

**Review by damianloew — COMMENTED** [review]
LGTM, could you please rebase changes and re-run CI on this target ~10 times before merge to make sure that there are no extra issues?

**Review by badochov — APPROVED** [review]
LGTM

**Review by damianloew — APPROVED** [review]

**Pull request merged**
